File: p3.py
#Time: O(2N)
#Space: O(1)
class Solution:
    def partitionLabels(self, S: str) -> List[int]:
        
        res=[]
        
        if S is None or len(S)==0:
            return res
        
        last = {c:i for i,c in enumerate(S)}
        
        maxi=0; size =1
        for i,c in enumerate(S):
            maxi = max(maxi,last[c])
            
            if i==maxi:
                res.append(size)
                size=1
            else:
                size+=1
        
        return res
            

File: p3.java
// Time Complexity : O(n)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :no


// Your code here along with comments explaining your approach

class Solution {
    public List<Integer> partitionLabels(String s) {
        HashMap<Character, Integer> map = new HashMap<>();
        //Store the last occurence of character
        for(int i=0; i< s.length(); i++){
            char c = s.charAt(i);
            map.put(c, i);
        }

        ArrayList<Integer> result = new ArrayList<>();
        int max = 0;
        int prev = 0;
        //When we reach the last occurence of a character in the string, put that in result
        for(int i=0; i< s.length(); i++){
            char c = s.charAt(i);
            int m = map.get(c);
            max = Math.max(max, m);
            if(max == i){
                result.add(max - prev + 1);
                prev = max+1;
            }
        }

        return result;
    }
}